### PR TITLE
policy-engine: gather and expose metrics

### PR DESCRIPTION
**Pull request by lucab**
This instruments policy-engine to gather some basic metrics (related
to graph serving to clients) and to expose them in Prometheus textual
format.

### CODE REVIEW

File: policy-engine/src/config.rs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 Alex Crawford
+// Copyright 2018 - Cincinnati authors.
```
Comment (crawford) [inline policy-engine/src/config.rs]: We can probably just drop these headers. Red Hat's policy is to omit them.

File: policy-engine/Cargo.toml [hunk 2]
```diff
@@ -4,13 +4,16 @@ version = "0.1.0"
...
+actix = "^0.7.6"
 actix-web = "^0.7.8"
 cincinnati = { path = "../cincinnati" }
 env_logger = "^0.5.10"
 failure = "^0.1.1"
 futures = "^0.1.23"
 hyper = "^0.12.6"
+lazy_static = "^1.2.0"
 log = "^0.4.3"
+prometheus = "^0.4.2"
```
Comment (crawford) [inline policy-engine/Cargo.toml:16]: Can you split this into a separate commit?

File: policy-engine/src/graph.rs [hunk 3]
```diff
@@ -18,20 +18,49 @@ use cincinnati::{Graph, CONTENT_TYPE};
...
+            HTTP_UPSTREAM_REQS.inc();
+            let timer = HTTP_SERVE_HIST.start_timer();
+            let serve = Client::new()
+                .request(ups_req)
                 .from_err::<Error>()
-                .and_then(|res| {
+                .map_err(|e| {
+                    HTTP_UPSTREAM_UNREACHABLE.inc();
+                    e
+                }).and_then(|res| {
```
Comment (crawford) [inline policy-engine/src/graph.rs]: Can you bump this down to the line below? This formatting is a bug in rustfmt that should have been fixed recently.

File: policy-engine/src/graph.rs [hunk 4]
```diff
@@ -40,16 +69,22 @@ pub fn index(req: HttpRequest<State>) -> Box<Future<Item = HttpResponse, Error =
...
-        ),
-        _ => Box::new(future::ok(HttpResponse::NotAcceptable().finish())),
+                }).then(move |r| {
+                    timer.observe_duration();
+                    r
+                });
+            Box::new(serve)
+        }
+        _ => {
+            HTTP_GRAPH_WRONG_REQS.inc();
```
Comment (crawford) [inline policy-engine/src/graph.rs]: I think `HTTP_GRAPH_BAD_REQS` is a more clear name for this metric.

File: policy-engine/src/main.rs [hunk 5]
```diff
@@ -45,18 +52,27 @@ fn main() -> Result<(), Error> {
...
         App::with_state(state.clone())
             .middleware(Logger::default())
             .route("/v1/graph", Method::GET, graph::index)
-    })
-    .bind((opts.address, opts.port))?
-    .run();
+    }).bind((opts.address, opts.port))?
+    .start();
+
+    sys.run();
```
Comment (crawford) [inline policy-engine/src/main.rs:67]: Does this end up waiting for both servers?

**Review by crawford — CHANGES_REQUESTED** [review]

File: policy-engine/src/config.rs (hunk 1 above)
Comment (lucab) [reply policy-engine/src/config.rs]: I assumed you meant the whole header (license et al.), so I completely dropped them all.

File: policy-engine/src/graph.rs (hunk 3 above)
Comment (lucab) [reply policy-engine/src/graph.rs]: Did a pass with nightly rustfmt, moved a few other things as well, it looks better now.

File: policy-engine/src/main.rs (hunk 5 above)
Comment (lucab) [reply policy-engine/src/main.rs:67]: Yes, this runs and waits for both servers (if one is panicking the other keep going).

**Comment by lucab** [comment]
Rebased and reformatted. On a server without any upstream, metrics would look like this:
```
$ curl -H "Accept: application/json" 127.0.0.1:8081/v1/graph ; curl 127.0.0.1:9081/metrics
# HELP http_graph_requests_total Total number of HTTP /v1/graph requests.
# TYPE http_graph_requests_total counter
http_graph_requests_total 1
# HELP http_graph_serve_duration_seconds HTTP graph serving latency in seconds.
# TYPE http_graph_serve_duration_seconds histogram
http_graph_serve_duration_seconds_bucket{le="0.005"} 1
http_graph_serve_duration_seconds_bucket{le="0.01"} 1
http_graph_serve_duration_seconds_bucket{le="0.025"} 1
http_graph_serve_duration_seconds_bucket{le="0.05"} 1
http_graph_serve_duration_seconds_bucket{le="0.1"} 1
http_graph_serve_duration_seconds_bucket{le="0.25"} 1
http_graph_serve_duration_seconds_bucket{le="0.5"} 1
http_graph_serve_duration_seconds_bucket{le="1"} 1
http_graph_serve_duration_seconds_bucket{le="2.5"} 1
http_graph_serve_duration_seconds_bucket{le="5"} 1
http_graph_serve_duration_seconds_bucket{le="10"} 1
http_graph_serve_duration_seconds_bucket{le="+Inf"} 1
http_graph_serve_duration_seconds_sum 0.00154805
http_graph_serve_duration_seconds_count 1
# HELP http_upstream_errors_total Total number of HTTP upstream unreachable errors.
# TYPE http_upstream_errors_total counter
http_upstream_errors_total 1
# HELP http_upstream_requests_total Total number of HTTP upstream requests.
# TYPE http_upstream_requests_total counter
http_upstream_requests_total 1
```

**Review by crawford — APPROVED** [review]
/lgtm
